### PR TITLE
Source Notion: Add Availability Strategy

### PR DESCRIPTION
**Pull request by ChristoGrab**
## What
Resolve open tickets for implementing [availability strategy](https://github.com/airbytehq/airbyte/issues/22263) and [oncall 2712](https://github.com/airbytehq/oncall/issues/2712).

## How
- Inherit the base HttpAvailabilityStrategy to handle skipping streams when encountering a 403 error.
- Added a slightly more Notion-specific message to output in the logs.
- Small change to handle cases where no next_cursor exists when checking for a next_page_token, avoiding a KeyError. Added to this PR as it was also affecting the unit test for the 403 handling.

## User Impact
Not a breaking change

## Note
- This PR will fail acceptance tests until [schema changes have been merged](https://github.com/airbytehq/airbyte/pull/30587). I'm submitting it now just to be ready.

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`7fc964a1da`](https://github.com/airbytehq/airbyte/pull/30750/commits/7fc964a1da07e8466198e3e2091ec5a3f3ae2714)) - ❌

⏲️  Total pipeline duration: 03mn28s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1695690611/7fc964a1da07e8466198e3e2091ec5a3f3ae2714/source-notion/1.1.3/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:7fc964a1da07e8466198e3e2091ec5a3f3ae2714)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`e89c410a77`](https://github.com/airbytehq/airbyte/pull/30750/commits/e89c410a77e1ad0ea9428fb56f25329fb0afafbf)) - ❌

⏲️  Total pipeline duration: 05mn11s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ❌        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1695792445/e89c410a77e1ad0ea9428fb56f25329fb0afafbf/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:e89c410a77e1ad0ea9428fb56f25329fb0afafbf)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`2b49e8d46e`](https://github.com/airbytehq/airbyte/pull/30750/commits/2b49e8d46ea539dbfd4b698481988f726d871b3c)) - ❌

⏲️  Total pipeline duration: 05mn00s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1695793709/2b49e8d46ea539dbfd4b698481988f726d871b3c/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2b49e8d46ea539dbfd4b698481988f726d871b3c)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

File: airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py [hunk 1]
```diff
@@ -167,3 +181,43 @@ def test_user_stream_handles_pagination_correctly(requests_mock):
...
+    (Blocks, Pages, 403, False, "This is likely due to insufficient permissions for your Notion integration. "),
+    (Users, None, 200, True, None)
+])
+def test_403_error_handling(stream, parent, status_code, expected_availability, expected_reason_substring):
+    """
+    Test that availability strategy flags streams with 403 error as unavailable
+    and returns custom Notion integration message.
+    """
+
+    with patch(f'source_notion.streams.{stream.__name__}._send_request') as mock_send_request:
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:197]: I suggest you use `requests_mock` for cases like this. It is more readable and does not require patching magic methods. See example [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-gitlab/unit_tests/test_streams.py#L153)

File: airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py [hunk 2]
```diff
@@ -167,3 +181,43 @@ def test_user_stream_handles_pagination_correctly(requests_mock):
...
+        mock_resp = requests.Response()
+        mock_resp.status_code = status_code
+
+        if status_code == 403:
+            mock_resp._content = b'{"object": "error", "status": 403, "code": "restricted_resource"}'
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:202]: I think it's worth moving the response content to the parameters of the test along with the status code and others

**Review by davydov-d — APPROVED** [review]
Minor comments regarding unit tests, otherwise looks good

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`5336f8bdc7`](https://github.com/airbytehq/airbyte/pull/30750/commits/5336f8bdc75b6a8f8013a3ab45a572cd6801fe9c)) - ❌

⏲️  Total pipeline duration: 04mn47s 

| Step                                                                 | Result   |
|:---------------------------------------------------------------------|:---------|
| Connector package install                                            | ✅        |
| Build source-notion docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                           | ✅        |
| Acceptance tests                                                     | ❌        |
| Code format checks                                                   | ✅        |
| Validate airbyte-integrations/connectors/source-notion/metadata.yaml | ✅        |
| Connector version semver check                                       | ✅        |
| Connector version increment check                                    | ✅        |
| QA checks                                                            | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1695830237/5336f8bdc75b6a8f8013a3ab45a572cd6801fe9c/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:5336f8bdc75b6a8f8013a3ab45a572cd6801fe9c)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

File: airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py (hunk 1 above)
Comment (ChristoGrab) [reply airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py]: Gotcha, thanks for sharing the example. I've refactored the test as suggested 👍

Comment (davydov-d) [reply airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py]: Thanks! Let's ship it

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`62ecaee3de`](https://github.com/airbytehq/airbyte/pull/30750/commits/62ecaee3ded623c031c1a3a9107637cd87b4893a)) - ❌

⏲️  Total pipeline duration: 04mn19s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-notion docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-notion                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ❌        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1696855664/62ecaee3ded623c031c1a3a9107637cd87b4893a/source-notion/1.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:62ecaee3ded623c031c1a3a9107637cd87b4893a)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-notion test report (commit [`2f310a96db`](https://github.com/airbytehq/airbyte/pull/30750/commits/2f310a96db88f4e35ddf095287be55203861146c)) - ✅

⏲️  Total pipeline duration: 04mn15s 

| Step                                                          | Result   |
|:--------------------------------------------------------------|:---------|
| Connector package install                                     | ✅        |
| Build source-notion docker image for platform(s) linux/x86_64 | ✅        |
| Unit tests                                                    | ✅        |
| Acceptance tests                                              | ✅        |
| Code format checks                                            | ✅        |
| Validate metadata for source-notion                           | ✅        |
| Connector version semver check                                | ✅        |
| Connector version increment check                             | ✅        |
| QA checks                                                     | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/christo_notion-availability-strategy/1696857092/2f310a96db88f4e35ddf095287be55203861146c/source-notion/1.2.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:2f310a96db88f4e35ddf095287be55203861146c)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-notion test
```

**Pull request merged**